### PR TITLE
Kafka-Tutorials: KTable Foreign Key Joins Tutorial

### PR DESCRIPTION
**Pull request by bbejeck**
This PR is my first attempt at a Kafka Tutorial (issue #193) and contains mostly all-new content.
The only existing files changed are:

1. `index.html`
2. `.semaphore/semaphore.yml`
3. `_data/tutorials.yml`

I validated the tutorial by running all the steps locally.

### CODE REVIEW

File: _includes/tutorials/fk-joins/kstreams/code/tutorial-steps/dev/console-producer-albums.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+docker exec -i schema-registry /usr/bin/kafka-avro-console-producer --topic albums --broker-list broker:9092\
+  --property "parse.key=true"\
```
Comment (bbejeck) [inline _includes/tutorials/fk-joins/kstreams/code/tutorial-steps/dev/console-producer-albums.sh]: as of the 5.4 release, the `kafka-avro-console-producer` accepts a `key-schema` parameter.  I've validated this works locally.

File: _includes/tutorials/fk-joins/kstreams/code/tutorial-steps/dev/track-purchases.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+100:{"id": 100, "album_id": 5, "song_title": "Houses Of The Holy", "price": 0.99}
```
Comment (bbejeck) [inline _includes/tutorials/fk-joins/kstreams/code/tutorial-steps/dev/track-purchases.json]: This is the correct format, it allows for passing in keys to `kafka-avro-console-producer`

File: _includes/tutorials/fk-joins/kstreams/code/tutorial-steps/dev/albums.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+5:{"id": 5, "title": "Physical Graffiti", "artist": "Led Zeppelin", "genre": "Rock"}
+6:{"id": 6, "title": "Highway to Hell",   "artist": "AC/DC", "genre": "Rock"}
```
Comment (bbejeck) [inline _includes/tutorials/fk-joins/kstreams/code/tutorial-steps/dev/albums.json]: This is the correct format, it allows for passing in keys to `kafka-avro-console-producer`

File: _includes/tutorials/fk-joins/kstreams/code/tutorial-steps/dev/wait-for-containers.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+echo "Waiting for the broker to become available ..."
+
+readiness_probe
+
+while [[ $? != 0 ]]; do
+    sleep 5
+    readiness_probe
+done
```
Comment (rspurgeon) [inline _includes/tutorials/fk-joins/kstreams/code/tutorial-steps/dev/wait-for-containers.sh]: There is a built in bash function (until) for this if you're interested...

```
until readiness_probe
do
  sleep 5
done

File: _includes/tutorials/fk-joins/kstreams/markup/dev/start-compose.adoc [hunk 5]
```diff
@@ -0,0 +1,5 @@
+And launch it by running:
+
++++++
+<pre class="snippet"><code class="shell">{% include_raw tutorials/filtering/kstreams/code/tutorial-steps/dev/docker-compose-up.sh %}</code></pre>
```
Comment (rspurgeon) [inline _includes/tutorials/fk-joins/kstreams/markup/dev/start-compose.adoc]: copy and paste moster...  `filtering` should be `fk-joins`

File: _includes/tutorials/fk-joins/kstreams/markup/dev/make-avro-dir.adoc [hunk 6]
```diff
@@ -0,0 +1,9 @@
+This tutorial uses three streams: one called `albums` that holds album reference data, one called `trackPurchases` that holds an update-stream of inbound music track purchases, and one called `musicInterestTable` that holds the result of a foreign-key join between purchases and albums.  In the case the inbound keys are different, but the `trackPurcases` stream has the id of the album in its value, so we can use the `KTable` foreign-key join funcitionality to extract the album id and perform the join.
```
Comment (rspurgeon) [inline _includes/tutorials/fk-joins/kstreams/markup/dev/make-avro-dir.adoc]: typo in `trackPurcases`

Comment (rspurgeon) [reply _includes/tutorials/fk-joins/kstreams/markup/dev/make-avro-dir.adoc]: > result of a foreign-key join between purchases and albums
Should this be more explicit and say... 'between trackPurchases and albums'?

File: _includes/tutorials/fk-joins/kstreams/markup/dev/make-joiner.adoc [hunk 7]
```diff
@@ -0,0 +1,9 @@
+For the ValueJoiner class, create the following file at `src/main/java/io/confluent/developer/MusicInterestJoiner.java`.
+
+When you join two tables in a relational database, by default you get a new table containing all of the columns of the left table plus all of the columns of the right table. When you join a table and a table, you get a new table, but you must be explicit about the value of that table—the combination between the value in the table and the assocaited value in the table. The `https://kafka.apache.org/{{ site.ak_javadoc_version }}/javadoc/org/apache/kafka/streams/kstream/ValueJoiner.html[ValueJoiner]` interface in the Streams API does this work. The single `apply()` method takes the "left" table and the "right" table values as parameters, and returns the value of the joined table as output. (Their keys are not a part of the equation, because they are equal by definition and do not change in the result.) As you can see here, this is just a matter of creating a `MusicInterest` object and populating it with the relevant fields of the input album and track purchase.
```
Comment (rspurgeon) [inline _includes/tutorials/fk-joins/kstreams/markup/dev/make-joiner.adoc]: > When you join two tables in a relational database, by default you get a new table containing all of the columns of the left table plus all of the columns of the right table. When you join a table and a table, you get a new table, but you must be explicit about the value of that table—the combination between the value in the table and the assocaited value in the table.

This was difficult to follow, maybe try a revision?

Comment (bbejeck) [reply _includes/tutorials/fk-joins/kstreams/markup/dev/make-joiner.adoc]: ack

File: _includes/tutorials/fk-joins/kstreams/markup/dev/make-avro-dir.adoc (hunk 6 above)
Comment (bbejeck) [reply _includes/tutorials/fk-joins/kstreams/markup/dev/make-avro-dir.adoc]: ack

File: _includes/tutorials/fk-joins/kstreams/markup/dev/start-compose.adoc (hunk 5 above)
Comment (bbejeck) [reply _includes/tutorials/fk-joins/kstreams/markup/dev/start-compose.adoc]: ack, thanks for that.

File: _includes/tutorials/fk-joins/kstreams/code/tutorial-steps/dev/wait-for-containers.sh (hunk 4 above)
Comment (bbejeck) [reply _includes/tutorials/fk-joins/kstreams/code/tutorial-steps/dev/wait-for-containers.sh]: thanks but I used pre-existing a `wait-for-containers.sh` file.  It seems most/all of the streams examples use the same format.  IMHO if we make that change, we can do a follow-up PR to change all files to use the same format.

**Comment by bbejeck** [comment]
the ci build failed on an unrelated tutorial `splitting`, but the test build for the KTable FK join passed.  Triggering a rebuild.

**Comment by bbejeck** [comment]
@rspurgeon I've updated per comments

**Comment by bbejeck** [comment]
`KTable FK Joins` build passed, failure due to unrelated tutorial failures
`KStream splitting`, `KSQL tumbling window`, and `KStreams final result`

**Comment by bbejeck** [comment]
The relevant test for this PR passed https://confluentinc.semaphoreci.com/jobs/8a995773-7b62-4f24-ad1b-eec6afc934f7

ping @colinhicks for review

File: _data/tutorials.yaml [hunk 8]
```diff
@@ -72,6 +72,15 @@ joining-stream-stream:
     ksql: enabled
     kstreams: enabled
 
+fk-joins:
+  title: "How to join a table and a table with a foriegn key"
+  meta-description: "Learn how to use the KTable foriegn key join functionality"
+  slug: "/table-table-foreign-key-joins"
+  problem: "you need to join two tables, but the primary keys are different so you need to use a foriegn key to complete the join"
+  introduction: "Suppose you are running an internet streaming music service where you offer albums or individual music tracks for sale.  You'd like to track trends in listener preference by joining the track purchases against the table of albums.  The issue is that the track purchase key doesn't align with the primary key for the album table.  But the value of the track purchase contains the id of the album.  Using the KTable-KTable foreign key join functionality, you can extract the album id from the track purchase and complete the join against the album table."
```
[the remark this answers is not in the document]
Comment (bbejeck) [reply _data/tutorials.yaml]: ack I'll take a look

File: _includes/tutorials/fk-joins/kstreams/code/configuration/dev.properties [hunk 9]
```diff
@@ -0,0 +1,16 @@
+application.id=-fk-joining-app
```
[the remark this answers is not in the document]
Comment (bbejeck) [reply _includes/tutorials/fk-joins/kstreams/code/configuration/dev.properties]: typo, also in `test.properties` fixed now.

**Comment by bbejeck** [comment]
the relevant test `KStreams table to table foreign key join tests` passed

**Comment by bbejeck** [comment]
@colinhicks updated this per comments

File: _includes/tutorials/fk-joins/kstreams/markup/test/make-test-file.adoc [hunk 10]
```diff
@@ -0,0 +1,8 @@
+First, create a test file at `configuration/test.properties`:
+
++++++
+<pre class="snippet"><code class="shell">{% include_raw tutorials/fk-joins/kstreams/code/configuration/test.properties %}</code></pre>
++++++
+
+You should take note of the `schema.registry.url` configuration.  The config is using a special pseudo-protocol `mock://..` which means our test
```
Comment (bbejeck) [inline _includes/tutorials/fk-joins/kstreams/markup/test/make-test-file.adoc]: just added this bit text as it explains the test configuration for schema registry i.e. using `MockSchemaRegistry`


\cc @colinhicks @rspurgeon

**Pull request merged**